### DISCUSSION TITLE
Hard coded context

### DISCUSSION THREAD

**Discussion by CommanderB** [ybizeul/StreamDeckWS#10] [General]
I have a need to change a button image based on an MQTT command. Traffic lights - red, amber, green. The command comes in on the MQTT listener, a switch and then set one button to the color and the other two buttons to an off image. The challenge I have is the context. The Red/Amber/Green change nodes have the context of their button (captured from a press to debug one time) hard coded. If I move or change one of the buttons, the context changes and breaks the flow.

So is there a better way of doing this? I tried with settings.id but unsuccessfully (I read that and switch on it for other flows).

![NR StreamDeck](https://user-images.githubusercontent.com/18680188/104781998-862d1100-5751-11eb-85dd-8ae377c3381a.png)

**Comment by ybizeul** [comment]
There is probably a better way. First one is to try and use the undocumented nodes `sd-in`/`sd-out` from `node-red-contrib-streamdeck-ws`, it's actually a node-red module created for easier integration with StreamDeckWS.

This module has the particularity to maintain a cache of all the keys sent to node-red. Here is what the `out` module looks like, but you will probably need to put the `in` module somewhere just for the purpose of maintaining the internal cache.

![image](https://user-images.githubusercontent.com/182035/104817910-ab3e8400-5824-11eb-8227-165e2ce50975.png)

You can set any `id` here, that's actually the same `id` as in StreamDeckWS on your computer, you don't have to fiddle with the context, as long as it has been seen once and sent through `sd-in`, it'll be in the cache (if anything else, from the last `willAppear` event when StreamDeck started)

The other way to do this if you don't want to use that module, is to use a flow variable to maintain a cache as a mapping of `id`/`context`, as they come is when StreamDeck displays it in the `willAppear` event. The same idea apply, you would have to attach the websocket entrypoint to that special function.

Does that make sense ?

**Reply by CommanderB** [reply]
Yes, thanks, I will try the node-red module - I wasnt aware of it. 

For the second option I would have to create my own flow that captures all the contexts from wllAppear events and store them in variables. Your module will make it easier :-)

**Comment by CommanderB** [comment]
Greatly simplified - thanks for the help
![SD TL](https://user-images.githubusercontent.com/18680188/104857063-05524e80-58e4-11eb-9a55-6f1c90a0dbeb.png)
